src/components/CharacterAssignment.jsx

```diff
@@ -0,0 +1,229 @@
+import React, { useEffect, useState } from 'react';
+import { DragDropContext } from 'react-beautiful-dnd';
+import ActorColumn from './ActorColumn.jsx'
+import UnassignedColumn from './UnassignedColumn';
+
+// Entire Drag and Drop context for assigning characters to actors
+const CharacterAssignment = ({actors, currentScript}) => {
+  const [columnOrder, setColumnOrder] = useState([])
+  const [columns, setColumns] = useState({
+    'unassignedCharacters': {
+      id:'unassignedCharacters',
+        title: 'Unassigned Characters',
+        characterIds: [],
+        lineCount: 0
+    }});
+  const [characters, setCharacters] = useState()
+
+  // Create the actor columns initially and edit columns when actors are added or removed
+  useEffect(() => {createOrUpdateActorColumns(actors)}, [actors]);
+
+  // get the character data on the initial load and when the script option changes
+  useEffect(() => { (async () => { // useEffect cannot take an async function. Must wrap async in regular function
+    if (currentScript === null) {
+      return;
+    }
+    try {
+      const response = await fetch('/api/script/' + currentScript + '/characters');
+
+      if (response.ok) {
+        const fetchedCharacters = await response.json()
+        resetActors()
+        setCharacters(fetchedCharacters);
+        setAssignedCharacters(fetchedCharacters);
+
+      } else {
+        console.error(`server error: ${response.body} when fetching character data`);
+      }
+    } catch (error) {
+      console.error(`network error: ${error} when fetching character data`);
+    }
+  })()}, [currentScript]);
+
+
+
+  // Reset the assigned characters to the actors so we don't have old character Ids interfering
+  const resetActors = () => {
+    const newColumns = {...columns}
+    for (let key in newColumns) {
+      newColumns[key].characterIds = [];
+      newColumns[key].lineCount = 0;
+    }
+    setColumns({...newColumns})
+  }
+
+
+  // Set all characters in their assigned column based on the fetched data
+  const setAssignedCharacters = (fetchedCharacters) => {
+    const newColumns = {...columns}
+
+    for (let key in fetchedCharacters){
+      let assignedActor = fetchedCharacters[key].actorId
+      if (assignedActor) {
+        newColumns[assignedActor].characterIds.push(fetchedCharacters[key].id)
+        newColumns[assignedActor].lineCount += fetchedCharacters[key].lineCount
+      } else {
+        newColumns.unassignedCharacters.characterIds.push(fetchedCharacters[key].id)
+        newColumns.unassignedCharacters.lineCount += fetchedCharacters[key].lineCount
+      }
+    }
+    setColumns({...newColumns})
+  }
+
+  // Create and update the columns for each of the actors and an Unassigned Characters column
+  const createOrUpdateActorColumns = (actors) => {
```

Comment:
Will this fail to create the columns if _no actors_ have been created yet?

Reply:
Good catch. It was still creating the unassigned characters column but once you added an actor it cleared out the unassigned characters column and generally freaked out. It's fixed now.